Initial (Github Actions) Workflows for Ansible Community Package release

anweshadas opened the pull request:
## Introduction 

The Release Manager(s) (RM) for Ansible Community Package, are always the Ansible Community Team members, who are Red Hatters. But with this endeavor we are trying to open up the release process to the community members. So an interested and designated group of people can be Release Ansible.
Who can be the release managers for Ansible Community Package?

Anyone from the Ansible Community can be the RM for Ansible Community Package. We will be forming a Release Manager Working Group from the community who will be not only doing the release but also taking the decisions and responsible for work in and around the release process (barring certain reasonable restrictions and limitations). 

## Automation of the current release process

We are trying to automate the current release process with Github Actions (GHA). We wrote the initial workflow(s). This workflow tries to follow the current release process as much as possible. 

#### Step 1 


Run the [ansible-release workflow] to create the release PR to the ansible-build-data repo.

[ansible-release workflow]:
https://github.com/anweshadas/ansible-build-data/blob/test/.github/workflows/ansible-release.yml

#### Step 2

After the PR is merged by a maintainer, run [upload-to-pypi] to upload the [distribution packages] to PyPI.


[upload-to-pypi]:
https://github.com/anweshadas/ansible-build-data/blob/test/.github/workflows/upload-to-pypi.yml

#### Step 3

After uploading the source to PyPI run [docsbuild-release workflow] to  create PR to ansible-documentation.

[docsbuild-release workflow]:
https://github.com/anweshadas/ansible-documentation/blob/test/.github/workflows/docsbuild-release.yaml

### Plan for betterment this workflow 

Here is the plan to improve the intial workflow

The process needs to be divided in two parts: Pre-Release and Release

Pre- Release includes:

- Creating PR in ansible-build-data
- Merging (manually) the above mentioned PR

Pre-requisite : A new ansible-core release will trigger the Pre-Release workflow

Release workflow includes:

- Make and upload GHA artifact
- Creating PR in ansible-documentation
- Releasing the tar in test.pypi.org
- Releasing the tar to PyPI
- Creating the Git Tag
- Releasing the [distribution packages] to PyPI

Pre-requisite : The release workflow will run (manually) only after the merging of ansible-build-data PR.

[distribution packages]:
https://packaging.python.org/en/latest/glossary/#term-Distribution-Package

The new upstream recommendations for building and publishing from github action to (Test)PyPI:
https://github.com/pypa/packaging.python.org/pull/1261

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+      - name: Install dependencies
+        working-directory: antsibull
+        run: |
+          python3 -m pip install --upgrade pip
+          python3 -m pip install ansible-core . ../antsibull-core ../antsibull-changelog
```
The release workflow should install antsibull from PyPI:

```suggestion
          python3 -m pip install --upgrade pip
          python3 -m pip install ansible-core antsibull
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+      - name: Install dependencies
+        working-directory: antsibull
+        run: |
+          python3 -m pip install --upgrade pip
+          python3 -m pip install ansible-core . ../antsibull-core ../antsibull-changelog
```
See comment above.

```suggestion
          python3 -m pip install --upgrade pip
          python3 -m pip install ansible-core antsibull
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+
+      - name: Create sdist dir using the deps file created in the ansible-build-data Release PR
+        run: |
+          antsibull-build rebuild-single "${{ inputs.ansible_version }}" --data-dir "${BUILD_DATA_DIR}" --build-file "${BUILDFILE}" --deps-file "${DEPSFILE}" --sdist-dir antsibull/sdist --debian
```
Hmm, I don't like this. Ideally, the previous job would save the artifacts that are generated by the release playbook and then this would simply upload them. As we stand, this is missing `--tags-file` and the `--build-file` and `--deps-file` args are redundant.

gotmax23 replied:
Also, the `--sdist-dir` should probably be `antsibull/dist`. We also build wheels.

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+        run: |
+          antsibull-build rebuild-single "${{ inputs.ansible_version }}" --data-dir "${BUILD_DATA_DIR}" --build-file "${BUILDFILE}" --deps-file "${DEPSFILE}" --sdist-dir antsibull/sdist --debian
+
+      - name: Upload Ansible sdist to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          repository-url: https://test.pypi.org/legacy/
+          packages-dir: antsibull/sdist
```
What's the idea here? This is not how test PyPI is supposed to be used. Why not have this upload the artifacts to actual PyPI?

webknjaz replied:
it's okay to have it if nightly runs are added, I think, to make sure that the pipeline is healthy. the testpypi job would have an unprotected environment, while the real pypi would require a human approval to start the job. but yes, it wouldn't be reasonable to have it in the same job for security reasons (privilege leak) and for structural reasons too (only one environment can be active in a job, meaning that two environments representing two different targets and having different protections set up would need a job each).

anweshadas replied:
First we need to test against test.pypi.org, then in the PR we can move to the production.

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+          -e antsibull_data_reset=false \
+          -e 'antsibull_ansible_version="${{ inputs.ansible_version }}"' \
+          -e 'antsibull_data_dir="{{ antsibull_data_git_dir }}/${{ inputs.ansible_major_version }}"' \
+          -e 'antsibull_build_file="ansible-${{ inputs.ansible_major_version }}.build"'
```
`antsibull_build_file` should not need to be manually specified.

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+          ref: main
+          path: antsibull
+
+      - name: Check out antsibull-core
+        uses: actions/checkout@v3
+        with:
+          repository: ansible-community/antsibull-core
+          ref: main
+          path: antsibull-core
+
+      - name: Check out antsibull-changelog
+        uses: actions/checkout@v3
+        with:
+          repository: ansible-community/antsibull-changelog
+          ref: main
+          path: antsibull-changelog
```
These don't need to be checked out if we're just using the playbook from antsibull.

```suggestion
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+          ref: main
+          path: antsibull
+
+      - name: Check out antsibull-core
+        uses: actions/checkout@v3
+        with:
+          repository: ansible-community/antsibull-core
+          ref: main
+          path: antsibull-core
+
+      - name: Check out antsibull-changelog
+        uses: actions/checkout@v3
+        with:
+          repository: ansible-community/antsibull-changelog
+          ref: main
+          path: antsibull-changelog
```
See the comment above.

``` suggestion
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+
+      - name: Pre-create sdist directory
+        run: |
+          mkdir -p antsibull/build
+          mkdir -p antsibull/sdist
```
```suggestion
          mkdir antsibull/build
          mkdir antsibull/sdist
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+on:
+  workflow_dispatch:
+    inputs:
+      ansible_version:
+        description: 'Release Version. Example : 11.1.0'
+        required: true
+      ansible_major_version:
+        description: 'Exmaple 11'
+        required: true
```
Future enhancement: make the workflow determine `ansible_version` from `ansible_major_version`,

webknjaz replied:
Maybe vice versa?

gotmax23 replied:
Oops, that's what I meant in the first place.

webknjaz commented on .github/workflows/ansible-release.yml:
Global permissions would better be empty and only jobs that need these elevated ones should declare them.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+        required: true
+env:
+  CI_COMMIT_MESSAGE: 'Ansible "${{ inputs.ansible_version }}": Dependencies, changelog and porting guide'
+  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
LOL I commented the same thing on the sibling PR, I guess many of those comments apply here too.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+        description: 'Release Version. Example : 11.1.0'
+        required: true
+      ansible_major_version:
+        description: 'Exmaple 11'
```
```suggestion
        description: Example 11
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+  workflow_dispatch:
+    inputs:
+      ansible_version:
+        description: 'Release Version. Example : 11.1.0'
```
```suggestion
        description: >-
          Release Version. Example: 11.1.0
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+on:
+  workflow_dispatch:
+    inputs:
+      ansible_version:
```
```suggestion
      ansible-version:
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+
+jobs:
+  build:
+    name: 'Build Ansible community distribution (${{ inputs.ansible_version }})'
```
Too long to fit in the UI

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+        uses: actions/checkout@v3
+        with:
+          path: antsibull/build/ansible-build-data
+          ref: main
```
Why is `ref` needed? Pretty sure this would be the default anyway.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+      - name: Install dependencies
+        working-directory: antsibull
+        run: |
+          python3 -m pip install --upgrade pip
```
In general, upgrading pip isn't really needed.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+
+      - name: Checking out to a new branch & setting the user details
+        run: |
+          cd antsibull/build/ansible-build-data
+          git checkout -b "publish-${{ inputs.ansible_version }}"
+          git config --global user.name "${{github.actor}}"
+          git config --global user.email "${{github.actor}}@users.noreply.github.com"
```
Same comments as in the other PR:
1) the user
2) the working dir setting

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+
+      - name: Commit ansible-build-data and push the changes to github
+        run: |
+         cd "antsibull/build/ansible-build-data/${{ inputs.ansible_major_version }}"
+         git add *.rst *.build *.deps *.yaml *.in validate-tags-ignores
+         git commit -a -m "${{ env.CI_COMMIT_MESSAGE }}"
+         git push origin "publish-${{ inputs.ansible_version }}"
```
working dir / double quotes

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+        run: |
+         cd "antsibull/build/ansible-build-data/${{ inputs.ansible_major_version }}"
+         git add *.rst *.build *.deps *.yaml *.in validate-tags-ignores
+         git commit -a -m "${{ env.CI_COMMIT_MESSAGE }}"
```
`-a` doesn't seem to be necessary

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,109 @@
+
+      - name: Create PR to the ansible-build-data
+        run: |
+          cd antsibull/build/ansible-build-data
+          gh pr create -B main -H "publish-${{ inputs.ansible_version }}" --title "${{ env.CI_COMMIT_MESSAGE }}"  --body "${{ env.CI_COMMIT_MESSAGE }}"
```
same as in the other PR

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+on:
+  workflow_dispatch:
+    inputs:
+      ansible_version:
+        description: 'Release Version. Example : 11.1.0'
+        required: true
+      ansible_major_version:
+        description: 'Exmaple 11'
+        required: true
```
same typos, kebab-case

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+        run: |
+          antsibull-build rebuild-single "${{ inputs.ansible_version }}" --data-dir "${BUILD_DATA_DIR}" --build-file "${BUILDFILE}" --deps-file "${DEPSFILE}" --sdist-dir antsibull/sdist --debian
+
+      - name: Upload Ansible sdist to PyPI
```
this is not true. it also uploads wheels

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+name: Upload-ansible-tarball-pypi
```
this is not true. it builds both sdist and wheel

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+          ansible-galaxy collection install 'git+https://github.com/ansible-collections/community.general.git'
+
+
+      - name: Create sdist dir using the deps file created in the ansible-build-data Release PR
```
This is not true. The command generates a wheel and an sdist, not just sdist. This is explained @ https://github.com/ansible-community/antsibull/issues/539.

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+    runs-on: ubuntu-latest
+    environment:
+      name: release
+      url: https://test.pypi.org/project/ansible/
```
make it nicer
```suggestion
      url: https://test.pypi.org/project/ansible/${{ inputs.ansible_version }}
```

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+    name: 'Upload Ansible community distribution (${{ inputs.ansible_version }}) to PyPI'
+    runs-on: ubuntu-latest
+    environment:
+      name: release
```
use a more accurate name
```suggestion
      name: testpypi
```

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,94 @@
+      name: release
+      url: https://test.pypi.org/project/ansible/
+    permissions:
+      id-token: write
```
This privilege mustn't be available to any of the build steps, only to the upload one. It's unacceptable to have it exposed to arbitrarily injected scripts, security-wise. This means that splitting this into separate jobs would be a hard requirement for this PR.

anweshadas commented:
Still working on this, more things left to be updated.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,88 @@
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh pr create -B main -H "publish-${{ inputs.ansible-version }}" --title "${{ env.CI_COMMIT_MESSAGE }}"  --body "${{ env.CI_COMMIT_MESSAGE }}"
```
What are these all single-letter CLI args?

anweshadas replied:
`-B` - The branch into which you want your code merged
`-H ` - The branch that contains commits for the pull request

Here is an example PR created by this workflow : `https://github.com/anweshadas/ansible-build-data/pull/10`

anweshadas replied:
Edited it to `gh pr create --base main --head "publish-${{ inputs.ansible-version }}" --title "${{ env.CI_COMMIT_MESSAGE }}"  --body "${{ env.CI_COMMIT_MESSAGE }}"`

webknjaz replied:
Great, thanks! A typical convention for anything resembling bash scripts is using long option names simply because it reduces the mental overhead for the readers, contributing to better maintainability.

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,109 @@
+          ansible-playbook -vv playbooks/build-single-release.yaml -e 'antsibull_ansible-version=${{ inputs.ansible-version }}' -e antsibull_data_reset=false
+
+      - name: Upload artifact
+        uses: actions/upload-artifact@v3
```
There's a v4 now. It has breaking changes, but since it's the initial integration, it should be fine to just go with it.

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,109 @@
+
+jobs:
+  build:
+    name: Build and Upload sdist and wheel
```
Let's drop "upload", it's confusing to thing that it publishes to PyPI.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,88 @@
+
+
+jobs:
+  build:
```
What's the difference between the build job in this workflow and the other one? Why do we need two workflows here?

anweshadas replied:
There are two steps right now (in the current process).
- This job creates the .deps, .tags and .yaml files required for the release.
- After the (manual) merging of the PR created by the aforesaid steps we run the second job to build the source tarball and wheel, upload it to PyPI and create the git tag.

webknjaz replied:
@anweshadas does that sdist need to contain the merged PR contents or can it be built ahead of time?

anweshadas replied:
> @anweshadas does that sdist need to contain the merged PR contents or can it be built ahead of time?

The `sdist` should not be build prior to the merging of the PR as we verify the `deps` file during the PR review.

webknjaz replied:
Why?

webknjaz replied:
I mean, what's missing? Can't it be built anyway and discarded if the review result wasn't acceptable?

felixfontein replied:
The current version of this job is uploading both sdist and wheel.

webknjaz replied:
@felixfontein I think my initial concern was regarding building the artifacts (both), not uploading (which is to be a separate thing).

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,109 @@
+
+    - name: Upload Ansible sdist and wheel to PyPI
+      uses: pypa/gh-action-pypi-publish@release/v1
+      with:
+        repository-url: https://pypi.org/project/ansible
```
This is an incorrect value, it won't work, because that's not a repository URL but a URL of a project hosted in the index. Delete it, it's not needed at all. The default value contains the correct PyPI upload URL anyway.
```suggestion
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,86 @@
+        description: Example 11
+        required: true
+env:
+  CI_COMMIT_MESSAGE: 'Ansible ${{ inputs.ansible-version }}: Dependencies, changelog and porting guide'
```
```suggestion
  CI_COMMIT_MESSAGE: 'Ansible ${{ inputs.ansible-version }}: Dependencies, changelog and porting guide'
  ANSIBLE_VERSION: '${{ inputs.ansible-version }}'
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,86 @@
+      - name: Checking out to a new branch & setting the user details
+        working-directory: antsibull/build/ansible-build-data
+        run: |
+          git checkout -b "publish-${{ inputs.ansible-version }}"
```
```suggestion
          git checkout -b "publish-${ANSIBLE_VERSION}"
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,86 @@
+        run: |
+          ansible-playbook -vv playbooks/build-single-release.yaml \
+          -e antsibull_data_reset=false \
+          -e 'antsibull_ansible_version=${{ inputs.ansible-version }}' \
```
```suggestion
          -e "antsibull_ansible_version=${ANSIBLE_VERSION}" \
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,86 @@
+          ansible-playbook -vv playbooks/build-single-release.yaml \
+          -e antsibull_data_reset=false \
+          -e 'antsibull_ansible_version=${{ inputs.ansible-version }}' \
+          -e 'antsibull_data_dir={{ antsibull_data_git_dir }}/${{ inputs.ansible-major-version }}'
```
This is already the default.

```suggestion
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,86 @@
+
+      - name: Commit ansible-build-data and push the changes to github
+        working-directory: 'antsibull/build/ansible-build-data/${{ inputs.ansible-major-version }}'
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
You don't need the token to push. It's already set up by the checkout action.

```suggestion
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,86 @@
+        run: |
+         git add *.rst *.build *.deps *.yaml *.in validate-tags-ignores
+         git commit -m "${{ env.CI_COMMIT_MESSAGE }}"
+         git push origin "publish-${{ inputs.ansible-version }}"
```
```suggestion
         git push origin "publish-${ANSIBLE_VERSION}"
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,86 @@
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh pr create -B main -H "publish-${{ inputs.ansible-version }}" --title "${{ env.CI_COMMIT_MESSAGE }}"  --body "${{ env.CI_COMMIT_MESSAGE }}"
```
These arguments are redundant anyways, but if you want to keep them, you should use `--head-branch` and `--base-branch` instead of the short options.

```suggestion
          gh pr create --title "${CI_COMMIT_MESSAGE}"  --body "${CI_COMMIT_MESSAGE}"
```

anweshadas replied:
Replaced it with `gh pr create --base main --head "publish-${{ inputs.ansible-version }}" --title "${{ env.CI_COMMIT_MESSAGE }}"  --body "${{ env.CI_COMMIT_MESSAGE }}"`

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+env:
+  BUILD_DATA_DIR: "antsibull/build/ansible-build-data/${{ inputs.ansible-major-version }}"
+  BUILDFILE: "ansible-${{ inputs.ansible-major-version }}.build"
+  DEPSFILE: "ansible-${{ inputs.ansible-version }}.deps"
```
```suggestion
  DEPSFILE: "ansible-${{ inputs.ansible-version }}.deps"
  ANSIBLE_VERSION: "${{ inputs.ansible-version }}"
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+      - name: Create sdist and wheel using the deps file created in the ansible-build-data Release PR
+        working-directory: antsibull
+        run: |
+          ansible-playbook -vv playbooks/build-single-release.yaml -e 'antsibull_ansible-version=${{ inputs.ansible-version }}' -e antsibull_data_reset=false
```
```suggestion
          ansible-playbook -vv playbooks/build-single-release.yaml -e "antsibull_ansible-version=${ANSIBLE_VERSION}" -e antsibull_data_reset=false
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+    runs-on: ubuntu-latest
+    environment:
+      name: pypi
+      url: https://pypi.org/project/ansible/${{ inputs.ansible-version }}
```
Is this the eight environment URL?

webknjaz replied:
@gotmax23 what do you mean? The link looks correct to me.

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+  git-tag:
+    needs: publish
+    name: 'Creates git tag for Ansible (${{ inputs.ansible-version }})'
+    runs-on: ubuntu-latest
```
```suggestion
    runs-on: ubuntu-latest
    permissions:
      contents: write
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+        run: |
+          git config --global user.name "Github Actions"
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git tag -a ${{ inputs.ansible-version }} -m "Ansible ${{ inputs.ansible-version }}: Changelog, Porting Guide and Dependent Collection Details"
```
```suggestion
          git tag -a "${ANSIBLE_VERSION}" -m "Ansible ${ANSIBLE_VERSION}: Changelog, Porting Guide and Dependent Collection Details"
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git tag -a ${{ inputs.ansible-version }} -m "Ansible ${{ inputs.ansible-version }}: Changelog, Porting Guide and Dependent Collection Details"
+          git push origin ${{ inputs.ansible-version }}
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
```suggestion
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+          Release Version. Example : 11.1.0
+        required: true
+      ansible-major-version:
+        description: 'Example 11'
```
```suggestion
        description: 'Ansible major version. Example: 11'
```

gotmax23 commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,106 @@
+        required: true
+
+env:
+  BUILD_DATA_DIR: "antsibull/build/ansible-build-data/${{ inputs.ansible-major-version }}"
+  BUILDFILE: "ansible-${{ inputs.ansible-major-version }}.build"
+  DEPSFILE: "ansible-${{ inputs.ansible-version }}.deps"
```
Are these env vars used? It doesn't look like it.

anweshadas replied:
Yes, I was not using them Thank you for pointing it out. Removed it now.

anweshadas commented:
Pushed changes based on the feedback.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,84 @@
+      - name: Install dependencies
+        working-directory: antsibull
+        run: |
+          python3 -m pip install ansible-core antsibull
+          ansible-galaxy collection install community.general
```
Split these commands into separate steps: having outputs of several processes together is usually hard to follow. And the individual error codes aren't distinguishable either. This recommendation is similar to what we do in Ansible playbooks.

anweshadas replied:
Edited the as per the suggestion.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,84 @@
+        working-directory: antsibull/build/ansible-build-data
+        run: |
+          git checkout -b "publish-${ANSIBLE_VERSION}"
+          git config --global user.name "Github Actions"
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
```
Setting up Git is also a separate action, let's not mix it with making a new branch. Also, setting user-global config doesn't need changing the directory.

anweshadas replied:
Edited the as per the suggestion.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,84 @@
+
+      - name: Test building a release with the defaults
+        working-directory: antsibull
+        run: |
+          ansible-playbook -vv playbooks/build-single-release.yaml \
+          -e antsibull_data_reset=false \
+          -e "antsibull_ansible_version=${ANSIBLE_VERSION}" \
```
That last trailing `\` can have a weird side effect that hard to spot. Also, using native YAML syntax, it's possible to produce a single-line command when this YAML file is parsed. With this, you don't need to worry about trailing spaces either, because all the lines will be glued together with whitespaces.
```suggestion
        run: >-
          ansible-playbook -vv playbooks/build-single-release.yaml
          -e antsibull_data_reset=false
          -e "antsibull_ansible_version=${ANSIBLE_VERSION}"
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,84 @@
+          ref: main
+          path: antsibull
+
+      - name: Pre-create build directory
+        run: mkdir -p antsibull/build
```
Strictly speaking, `mkdir` isn't really necessary, because `git clone` always creates the target directory.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,84 @@
+        working-directory: antsibull/build/ansible-build-data
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh pr create --base main --head "publish-${{ inputs.ansible-version }}" --title "${{ env.CI_COMMIT_MESSAGE }}"  --body "${{ env.CI_COMMIT_MESSAGE }}"
```
Here, we can use the same "multi-line but really one-line" trick:
```suggestion
        run: >-
          gh pr create
          --base main
          --head "publish-${{ inputs.ansible-version }}"
          --title "${{ env.CI_COMMIT_MESSAGE }}"
          --body "${{ env.CI_COMMIT_MESSAGE }}"
```
isn't it cleaner when flags are paired with their values? I also like using `--<flag>=<value>` style for even better clarity, but that's up to you.

anweshadas replied:
Edited the as per the suggestion.

webknjaz replied:
Looks like you forgot to commit this?

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,84 @@
+name: ansible-release
```
No need to repeat the default value with is the filename. Normally, this field is for human-readable title that shows up in the GH UI.

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,104 @@
+          git config --global user.name "Github Actions"
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git tag -a "${ANSIBLE_VERSION}" -m "Ansible ${ANSIBLE_VERSION}: Changelog, Porting Guide and Dependent Collection Details"
+          git push origin ${{ inputs.ansible-version }}
```
Shouldn't this also use `"${ANSIBLE_VERSION}"`?

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,104 @@
+        run: |
+          git config --global user.name "Github Actions"
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git tag -a "${ANSIBLE_VERSION}" -m "Ansible ${ANSIBLE_VERSION}: Changelog, Porting Guide and Dependent Collection Details"
```
Looks like this tags whatever's the last commit on the main branch. So if there's a race condition with something that manages squeeze in after the PR merge but before running this workflow, some other commit may end up being tagged, not the immediate PR merge. To combat this, we should probably track the merge commit associated with the PR.

anweshadas replied:
Yes, this will be a future improvement.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,84 @@
+        description: >-
+          Release Version. Example : 11.1.0
+        required: true
+      ansible-major-version:
+        description: Example 11
+        required: true
```
This input isn't really needed. Instead, calculate the major version from `ansible-version` and set that as a step/job output for reuse in runtime.

webknjaz commented:
@anweshadas I see that you made 2 workflows with separate triggers and most of the process being the same. Do I understand correctly, that it's one process split in half with a pause/gate in the middle?

It looks like this doesn't really need a separate trigger. Instead, I'd make one main workflow with `workflow_dispatch` and put parts into reusable workflows with `workflow_call`. The reusable workflows would be "called" (or "included") from that main one. And the workflow run would include all the bits. The job happening after the PR creation would be marked with some environment (like `await-pr-merge`) and with required reviews for that environment (configured in the repository settings), it'd pause the process. Once the PR's merged, you'd click a button in the paused workflow run to make it proceed.

With that, you could even generate dists together with making the PR, save them as workflow artifacts and only perform upload+tagging after unpausing. And since it's a single trigger, you'd have to type in the version just once.

anweshadas commented:
> @anweshadas I see that you made 2 workflows with separate triggers and most of the process being the same. Do I understand correctly, that it's one process split in half with a pause/gate in the middle?


I will learn about this and about this in the upcoming PRs.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,90 @@
+        run: |
+          git checkout -b "publish-${ANSIBLE_VERSION}"
+
+      - name: Setting the user details
+        working-directory: antsibull/build/ansible-build-data
```
No need to chdir, as mentioned in https://github.com/ansible-community/ansible-build-data/pull/265/files#r1442368714
```suggestion
      - name: Configure user-global Git author and committer metadata
```

anweshadas replied:
Updated.

webknjaz commented on .github/workflows/upload-to-pypi.yml:
```diff
@@ -0,0 +1,104 @@
+      - name: Create sdist and wheel using the deps file created in the ansible-build-data Release PR
+        working-directory: antsibull
+        run: |
+          ansible-playbook -vv playbooks/build-single-release.yaml -e "antsibull_ansible-version=${ANSIBLE_VERSION}" -e antsibull_data_reset=false
```
typo?
```suggestion
          ansible-playbook -vv playbooks/build-single-release.yaml -e "antsibull_ansible_version=${ANSIBLE_VERSION}" -e antsibull_data_reset=false
```

anweshadas replied:
Yes, updated.

gotmax23 commented:
I agree with @webknjaz here. If possible, I'd really like to see the dists uploaded as GHA artifacts that can be reviewed and tested as part of the PR and then have those same artifacts uploaded to PyPI after the PR is merged instead of building them twice.

anweshadas commented:
@webknjaz @gotmax23 I updated the PR based on the review. These are the following changes:

1. Added a new job to pause the workflow, created a new environment: `pypirelease` for that .
2. Added `release-managers` as reviewers, could not add `ansible-community/release-management-wg`. (Will look into it why)
3. Instead of 2 workflows (a. creating the sdist & wheel and b. uploading it to PyPI & creating the git tag) we will have one workflow. If the changes look good  I will remove the other file (`upload-to-pypi.yml`).

gotmax23 commented:
> Added release-managers as reviewers, could not add ansible-community/release-management-wg. (Will look into it why)

You probably have to add that group as maintainers of this repo first.

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,152 @@
+        run: echo 'Review Done'
+
+  publish:
+    needs: [build, prreview]
+    name: 'Upload Ansible (${{ inputs.ansible-version }}) to PyPI'
+    runs-on: ubuntu-latest
+    environment:
+      name: pypi
+      url: https://pypi.org/project/ansible
```
Why not

```suggestion
    needs: [build]
    name: 'Upload Ansible (${{ inputs.ansible-version }}) to PyPI'
    runs-on: ubuntu-latest
    environment:
      name: pypirelease
      url: https://pypi.org/project/ansible
```

and then remove the extra job?

cybette commented:
Added `release-management-wg` team as maintainers

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+    name: 'Upload Ansible (${{ inputs.ansible-version }}) to PyPI'
+    runs-on: ubuntu-latest
+    environment:
+      name: pypirelease
```
Why not just
```suggestion
      name: pypi
```
?

A GitHub environment corresponds to the deployment target (a place — PyPI), not the process (release).

anweshadas replied:
Updated.

webknjaz replied:
@anweshadas thanks! Did you also update it in the repo settings?

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+  # publish job downloads the arifacts and publish it to PyPI
+
+  publish:
+    needs: [build]
```
Since it's just one job, let's not make it a list:
```suggestion
    needs: build
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+          gh pr create
+          --base main
+          --head "publish-${{ inputs.ansible-version }}"
+          --title "${{ env.CI_COMMIT_MESSAGE }}"
+          --body "${{ env.CI_COMMIT_MESSAGE }}"
```
I think it's confusing to use the same text as both title and description of the PR. Can we separate them by context/semantics?

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+    runs-on: ubuntu-latest
+    environment:
+      name: pypirelease
+      url: https://pypi.org/project/ansible
```
Since we know the version at this point, let's make the URL more accurate:
```suggestion
      url: https://pypi.org/project/ansible/${{ inputs.ansible-version }}
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+          gh pr create
+          --base main
+          --head "publish-${{ inputs.ansible-version }}"
+          --title "${{ env.CI_COMMIT_MESSAGE }}"
```
Perhaps
```suggestion
          --title "Release Ansible v${{ inputs.ansible-version }}"
```
would do.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+        required: true
+env:
+  CI_COMMIT_MESSAGE: 'Ansible ${{ inputs.ansible-version }}: Dependencies, changelog and porting guide'
+  ANSIBLE_VERSION: '${{ inputs.ansible-version }}'
```
```suggestion
  ANSIBLE_VERSION: ${{ inputs.ansible-version }}
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+
+jobs:
+  build:
+    name: 'Build Ansible (${{ inputs.ansible-version }})'
```
```suggestion
    name: Build Ansible (${{ inputs.ansible-version }})
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+
+  publish:
+    needs: [build]
+    name: 'Upload Ansible (${{ inputs.ansible-version }}) to PyPI'
```
```suggestion
    name: Upload Ansible (${{ inputs.ansible-version }}) to PyPI
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+
+  git-tag:
+    needs: publish
+    name: 'Creates git tag for Ansible (${{ inputs.ansible-version }})'
```
```suggestion
    name: Creates git tag for Ansible (${{ inputs.ansible-version }})
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+        description: Example 11
+        required: true
+env:
+  CI_COMMIT_MESSAGE: 'Ansible ${{ inputs.ansible-version }}: Dependencies, changelog and porting guide'
```
```suggestion
  CI_COMMIT_MESSAGE: >-
    Ansible ${{ inputs.ansible-version }}:
    Dependencies, changelog and porting guide
```

webknjaz reviewed (commented):
(nitpick) Let's the unnecessary YAML quotes that are added added inconsistently in arbitrary places.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+      - name: Set up Python 3.11
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
```
(one more)
```suggestion
          python-version: 3.11
```

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+    steps:
+
+      - name: Check out ansible-build-data
+        uses: actions/checkout@v3
```
I think, this needs an explicit `ref` input set. By default, it'll probably target the commit attached to the original `workflow_dispatch` event that triggered this workflow run. However, we need to tag the merge commit after the PR, not the commit before the PR, since we want the tagged version to contain changes from the PR, right?

webknjaz replied:
If you want to postpone this — it's fine. But then, you need to at least force this specific checkout step to fetch the top of the main branch, not some old pre-PR commit:
```suggestion
        uses: actions/checkout@v3
        with:
          ref: ${{ github.event.repository.default_branch }}
```
(and implement what I suggested above in a follow-up PR)

anweshadas replied:
> To implement this, we need to save the PR number as an output of the `build` job above. Then, we'd access that PR number in this job, and query GitHub API (or maybe just use GH CLI, not sure) to retrieve the PR metadata and see which commit exactly is the merge commit associated with that PR.


I will file a separate PR on this context as a future enhancement.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+      - name: Check out ansible-build-data
+        uses: actions/checkout@v3
+
+      - name: Create git tag
```
May I suggest also creating a GitHub Release? These tend to show up nicer on GH and are still associated with tags.

anweshadas replied:
This is also a new feature (enhancement), I will file a separate on this in future.

webknjaz replied:
Okay, fair. Though, it's just another interface to making a tag so it might be viewed as invoking a different command rather than adding a new feature. But that's up to you, of course.

webknjaz commented:
@anweshadas another review round complete. Part of the comments are style normalization — you can safely accept the suggested changes in a batch from the Files tab in this PR.
And there's a few comments pointing out more serious problems that I believe should be addressed before merging this PR.

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,150 @@
+        run: >-
+          gh pr create
+          --base main
+          --head "publish-${{ inputs.ansible-version }}"
+          --title "Release Ansible v${{ inputs.ansible-version }}"
+          --body "${{ env.CI_COMMIT_MESSAGE }}"
```
```suggestion
          --head "publish-${ANSIBLE_VERSION}"
          --title "Release Ansible v${ANSIBLE_VERSION}"
          --body "${CI_COMMIT_MESSAGE}"
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,150 @@
+          git config --global user.name "Github Actions"
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git tag -a "${ANSIBLE_VERSION}" -m "Ansible ${ANSIBLE_VERSION}: Changelog, Porting Guide and Dependent Collection Details"
+          git push origin ${ANSIBLE_VERSION}}
```
```suggestion
          git push origin "${ANSIBLE_VERSION}"
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,150 @@
+          git config --global user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git tag -a "${ANSIBLE_VERSION}" -m "Ansible ${ANSIBLE_VERSION}: Changelog, Porting Guide and Dependent Collection Details"
+          git push origin ${ANSIBLE_VERSION}}
+
```
```suggestion
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,150 @@
+        env:
+          # Make result better readable
+          ANSIBLE_CALLBACK_RESULT_FORMAT: yaml
+
```
```suggestion
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+        run: |
+          python3 -m pip install ansible-core
+          python3 -m pip install antsibull
+          ansible-galaxy collection install community.general
```
Actually, I think we already have one.

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,144 @@
+        run: |
+          python3 -m pip install ansible-core
+          python3 -m pip install antsibull
+          ansible-galaxy collection install community.general
```
```suggestion
          ansible-galaxy install -r requirements.yml
```

felixfontein commented:
We tried to apply most of the feedback. PTAL.

webknjaz commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,150 @@
+      - name: Building a release with the defaults
+        working-directory: antsibull
+        run: >-
+          ansible-playbook -vv playbooks/build-single-release.yaml
```
@gotmax23 so I have a comment for you which I believe might need to be implemented in `antsibull`. This playbook builds the dists but doesn't perform the metadata validation. That must happen early in the process. Specifically, calling `twine check --strict` must succeed. My `pypi-upload` action is invoked late in the process and it does perform a non-strict check, which may fail the upload last moment which would be unpleasant. The PyPI backend might also reject uploads because of the problems with dists, some of which (but not all) can be caught in the strict mode.
Running the check as early as possible (on merges to `main` and pull requests) reveals metadata problems in a way that they can be fixed before attempting to build though this workflow.

So this workflow is the last line of defense that may reveal problems before the PR is created. I don't know if there's currently other places where it would be possible to inject this check, but perhaps putting it into that playbook would be useful. Something that could abort a release with broken meta.

gotmax23 replied:
Yeah, adding it to the playbook (or _maybe_ shelling out to it from the python code) definitely makes sense to me. Can you open an issue in ansible's issue tracker?

webknjaz replied:
Done: #577. Marking this thread as resolved.

webknjaz reviewed (approved):
You have my blessing, I suppose..

gotmax23 commented:
Requesting a review from myself. I'll give this another look.

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,150 @@
+        working-directory: >-
+          antsibull/build/ansible-build-data/${{ inputs.ansible-major-version }}
+        run: |
+         git add *.rst *.build *.deps *.yaml *.in validate-tags-ignores
```
I would prefer a `git add .` here. I don't want to have to have to adjust this in the future if we add additional files.

felixfontein replied:
```suggestion
         git add .
```

gotmax23 commented on .github/workflows/ansible-release.yml:
```diff
@@ -0,0 +1,150 @@
+          antsibull/build/ansible-build-data/${{ inputs.ansible-major-version }}
+        run: |
+         git add *.rst *.build *.deps *.yaml *.in validate-tags-ignores
+         git commit -m "${{ env.CI_COMMIT_MESSAGE }}"
```
```suggestion
         git commit -m "${CI_COMMIT_MESSAGE}"
```

gotmax23 commented:
Other than that, this looks good to me. Thanks @anweshadas for all your work on this and for responding to all the feedback! This is really great.

felixfontein reviewed (approved):
Let's get this merged so we can do further improvements in follow-up PRs :)

felixfontein commented:
CI failures are unrelated.

felixfontein commented:
Thanks @anweshadas @gotmax23 @webknjaz for your work on this!
